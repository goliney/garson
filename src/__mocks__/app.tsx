import React from 'react';
import { render } from 'ink-testing-library';
import { Text } from 'ink';

const initialView = <Text>Initializing tests...</Text>;

export const app = render(initialView);

app.unmount = jest.fn().mockImplementation(() => app.rerender(initialView));
