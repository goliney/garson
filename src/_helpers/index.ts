// Components
export { Question } from './components/question';
export { Line } from './components/line';

// Hooks
export { useKeyHandler } from './use-key-handler';
export { useEnterKeyHandler } from './use-enter-key-handler';
