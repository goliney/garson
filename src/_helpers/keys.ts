export const ARROW_UP = '\u001B[A';
export const ARROW_DOWN = '\u001B[B';
export const ARROW_LEFT = '\u001B[D';
export const ARROW_RIGHT = '\u001B[C';
export const ENTER = '\r';
export const CTRL_C = '\x03';
export const BACKSPACE = '\x08';
export const DELETE = '\x7F';
export const SPACE = ' ';
