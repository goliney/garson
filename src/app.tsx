import { Box, render } from 'ink';
import React from 'react';

export const app = render(<Box>Initializing...</Box>, {
  debug: false,
});
