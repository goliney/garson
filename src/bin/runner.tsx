import React from 'react';
import { Box, render } from 'ink';
import { GarsonConfig } from '../types';

const { rerender, unmount } = render(<Box>Initializing...</Box>, {
  debug: false,
});

export function runner({ results, prompts, action }: GarsonConfig) {
  const [{ prompt, key }, ...rest] = prompts;
  rerender(
    prompt(result => {
      rerender(<React.Fragment />); // clear the previous prompt from the screen
      const newResults = { ...results, [key]: result };
      if (rest.length === 0) {
        // it was the last prompt in a chain
        const actionResponse = action(newResults);
        if (actionResponse) {
          runner(actionResponse);
        } else {
          unmount();
        }
      } else {
        // there is another prompt
        runner({ results: newResults, prompts: rest, action });
      }
    }),
  );
}
