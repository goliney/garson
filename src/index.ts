import { KeyType, PromptCallback, Results, ActionCallback, GarsonHandlerParams } from './types';

export function garson(results?: Results) {
  return garsonHandler({ results });
}

function garsonHandler({ results = {}, prompts = [] }: GarsonHandlerParams) {
  return {
    prompt: (key: KeyType, prompt: PromptCallback) =>
      garsonHandler({ results, prompts: [...prompts, { key, prompt }] }),
    action: (action: ActionCallback) => ({
      results,
      prompts,
      action,
    }),
  };
}
