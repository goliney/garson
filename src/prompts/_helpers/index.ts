// Components
export { Question } from './components/question';

// Misc
export { useKeyHandler } from './use-key-handler';
