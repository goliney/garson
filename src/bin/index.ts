import { runner } from './runner';

const cwd = process.cwd();
let garsonConfig;

try {
  // eslint-disable-next-line import/no-dynamic-require, global-require
  garsonConfig = require(`${cwd}/garson.config.js`);
} catch (e) {
  console.error('Failed to load config', e);
}

if (garsonConfig) {
  runner(garsonConfig);
}
