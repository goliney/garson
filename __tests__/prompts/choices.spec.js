import { garson, prompts } from '../../src';
import { runner } from '../../src/bin/runner';
import { app } from '../../src/app';
import { ARROW_DOWN, ARROW_UP, ENTER } from '../../src/_helpers/keys';
import { stripColorsFromLastFrame, write } from '../helpers';

jest.mock('../../src/app');

const actionSpy = jest.fn();

const choicesOptions = {
  message: 'What git command you want to run?',
  items: [
    { label: 'See current branch', value: 'git branch' },
    { label: 'Checkout to master', value: 'git checkout master' },
    { label: 'See status', value: 'git status' },
  ],
};

const config = garson()
  .prompt('command', prompts.choices(choicesOptions))
  .action(actionSpy);

const configWithNumericInput = garson()
  .prompt('command', prompts.choices({ ...choicesOptions, isNumericInputEnabled: true }))
  .action(actionSpy);

describe('Choices', () => {
  beforeEach(async () => {
    actionSpy.mockClear();
    runner(config);
    await new Promise(resolve => setTimeout(resolve));
  });

  afterEach(() => {
    app.unmount();
  });

  test('Initial render', () => {
    expect(stripColorsFromLastFrame()).toMatchSnapshot();
  });

  test('Arrow controls', async () => {
    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ Checkout to master');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ See status');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_DOWN);
    expect(stripColorsFromLastFrame()).toContain('▇ See status');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ Checkout to master');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ See current branch');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();

    await write(ARROW_UP);
    expect(stripColorsFromLastFrame()).toContain('▇ See current branch');
    expect(stripColorsFromLastFrame()).toMatchSnapshot();
  });

  test('Select', async () => {
    await write('3'); // since isNumericInputEnabled is false, nothing should happen
    expect(actionSpy).not.toHaveBeenCalled();

    await write(ARROW_DOWN);
    await write(ARROW_DOWN);
    await write(ENTER); // select third option
    expect(actionSpy).toHaveBeenCalledWith({ command: 'git status' });
  });

  describe('With numeric input enabled', () => {
    test('Selects correct item', async () => {
      app.unmount();
      runner(configWithNumericInput);
      await new Promise(resolve => setTimeout(resolve));
      expect(stripColorsFromLastFrame()).toContain('▇ 1. See current branch');
      expect(stripColorsFromLastFrame()).toMatchSnapshot();
      await write('4');
      expect(actionSpy).not.toHaveBeenCalled();
      await write('3');
      expect(actionSpy).toHaveBeenCalledWith({ command: 'git status' });
    });

    test('Checks the number of items', async () => {
      app.unmount();
      expect(() => {
        garson()
          .prompt(
            'command',
            prompts.choices({
              ...choicesOptions,
              items: new Array(10),
              isNumericInputEnabled: true,
            })
          )
          .action(actionSpy);
      }).toThrow('If isNumericInputEnabled is true, the length of choices must be less than 10');
    });
  });
});
