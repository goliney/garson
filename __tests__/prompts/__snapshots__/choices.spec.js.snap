// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choices Arrow controls 1`] = `
"[32m?[39m [32mWhat git command you want to run?[39m
  See current branch
[38;2;0;87;255mâ–‡[39m [38;2;0;87;255mCheckout to master[39m
  See status"
`;

exports[`Choices Arrow controls 2`] = `
"[32m?[39m [32mWhat git command you want to run?[39m
  See current branch
  Checkout to master
[38;2;0;87;255mâ–‡[39m [38;2;0;87;255mSee status[39m"
`;

exports[`Choices Arrow controls 3`] = `
"[32m?[39m [32mWhat git command you want to run?[39m
  See current branch
  Checkout to master
[38;2;0;87;255mâ–‡[39m [38;2;0;87;255mSee status[39m"
`;

exports[`Choices Arrow controls 4`] = `
"[32m?[39m [32mWhat git command you want to run?[39m
  See current branch
[38;2;0;87;255mâ–‡[39m [38;2;0;87;255mCheckout to master[39m
  See status"
`;

exports[`Choices Arrow controls 5`] = `
"[32m?[39m [32mWhat git command you want to run?[39m
[38;2;0;87;255mâ–‡[39m [38;2;0;87;255mSee current branch[39m
  Checkout to master
  See status"
`;

exports[`Choices Arrow controls 6`] = `
"[32m?[39m [32mWhat git command you want to run?[39m
[38;2;0;87;255mâ–‡[39m [38;2;0;87;255mSee current branch[39m
  Checkout to master
  See status"
`;

exports[`Choices Initial render 1`] = `
"[32m?[39m [32mWhat git command you want to run?[39m
[38;2;0;87;255mâ–‡[39m [38;2;0;87;255mSee current branch[39m
  Checkout to master
  See status"
`;

exports[`Choices With numeric input enabled Selects correct item 1`] = `
"[32m?[39m [32mWhat git command you want to run?[39m
[38;2;0;87;255mâ–‡[39m [38;2;0;87;255m1. [39m[38;2;0;87;255mSee current branch[39m
  2. Checkout to master
  3. See status"
`;
